### PR TITLE
TRUNK-4640 "Throwable.printStackTrace(...) should never be called"

### PR DESCRIPTION
**Pull request by tomaszmueller**

### CODE REVIEW

**Comment by tomaszmueller** [comment]
@rkorytkowski https://issues.openmrs.org/browse/TRUNK-4640
I'm not sure about the cases where printstacktrace to PrintWriter was used(in Format.java,  SchedulerUtil.java, HL7ServiceImpl.java, StartupErrorFilterModel.java)

File: api/src/main/java/org/openmrs/hl7/impl/HL7ServiceImpl.java [hunk 1]
```diff
@@ -902,9 +902,10 @@ private void setFatalError(HL7InQueue hl7InQueue, String error, Throwable cause)
...
 			StringWriter sw = new StringWriter();
-			PrintWriter pw = new PrintWriter(sw, true);
-			cause.printStackTrace(pw);
-			pw.flush();
+			//			PrintWriter pw = new PrintWriter(sw, true);
+			//			cause.printStackTrace(pw);
+			log.error(cause);
+			//			pw.flush();
 			sw.flush();
 			hl7InError.setErrorDetails(OpenmrsUtil.shortenedStackTrace(sw.toString()));
```
Comment (rkorytkowski) [inline api/src/main/java/org/openmrs/hl7/impl/HL7ServiceImpl.java]: Please use hl7InError.setErrorDetails(ExceptionUtils.getStackTrace(case)); instead and get rid off StringWriter entirely. It's fine to keep log.error(cause).

File: api/src/main/java/org/openmrs/scheduler/SchedulerUtil.java [hunk 2]
```diff
@@ -141,7 +141,8 @@ public static void sendSchedulerError(Throwable throwable) {
 	public static String getExceptionAsString(Throwable t) {
 		StringWriter stringWriter = new StringWriter();
 		PrintWriter printWriter = new PrintWriter(stringWriter, true);
-		t.printStackTrace(printWriter);
+		log.error(t);
+		//		t.printStackTrace(printWriter);
 		printWriter.flush();
 		stringWriter.flush();
 		return stringWriter.toString();
```
Comment (rkorytkowski) [inline api/src/main/java/org/openmrs/scheduler/SchedulerUtil.java]: This should not call log.error(t). It should return ExceptionUtil.getStrackTrace(t);

File: api/src/main/java/org/openmrs/util/Format.java [hunk 3]
```diff
@@ -79,8 +79,9 @@ public static String format(Date date, Locale locale, FORMAT_TYPE type) {
 	
 	public static String format(Throwable t) {
 		StringWriter sw = new StringWriter();
-		PrintWriter pw = new PrintWriter(sw);
-		t.printStackTrace(pw);
+//		PrintWriter pw = new PrintWriter(sw);
+//		t.printStackTrace(pw);
+        log.error(t);
 		return t + "\n" + sw.toString();
```
Comment (rkorytkowski) [inline api/src/main/java/org/openmrs/util/Format.java]: This should not call log.error(t);. It should return t + "\n" + ExceptionUtil.getStackTrace(t);

File: web/src/main/java/org/openmrs/web/filter/startuperror/StartupErrorFilterModel.java [hunk 4]
```diff
@@ -39,8 +43,9 @@ public StartupErrorFilterModel(Throwable t) {
 		errorAtStartup = t;
 		
 		StringWriter sw = new StringWriter();
-		PrintWriter pw = new PrintWriter(sw);
-		t.printStackTrace(pw);
+		//		PrintWriter pw = new PrintWriter(sw);
+		//        t.printStackTrace(pw);
+		log.error(t);
 		stacktrace = sw.toString();
```
Comment (rkorytkowski) [inline web/src/main/java/org/openmrs/web/filter/startuperror/StartupErrorFilterModel.java]: This should not call log.error(t). It should set stacktrace = ExceptionUtil.getStackTrace(t);.

**Comment by tomaszmueller** [comment]
@rkorytkowski done

**Comment by rkorytkowski** [comment]
well done!

**Pull request merged**